File: styx-js/src/ledger/vector-clock.js
// styx-js/src/ledger/vector-clock.js
// 2-element vector clock for the Styx 2-peer system

import { uint32BE, readUint32BE, concatBytes } from '../utils.js';

/** @enum {string} */
export const CausalRelation = {
  BEFORE: 'before',
  AFTER: 'after',
  CONCURRENT: 'concurrent',
  EQUAL: 'equal',
};

/**
 * 2-element vector clock for Styx peers A and B.
 * Immutable — all mutations return a new VectorClock.
 */
export class VectorClock {
  /**
   * @param {number} a - Counter for peer A
   * @param {number} b - Counter for peer B
   */
  constructor(a, b) {
    this.a = a;
    this.b = b;
    Object.freeze(this);
  }

  /** Zero vector clock */
  static zero() {
    return new VectorClock(0, 0);
  }

  /** From JSON { a, b } */
  static fromJSON(json) {
    return new VectorClock(json.a, json.b);
  }

  /** From 8-byte big-endian buffer */
  static fromBytes(bytes) {
    return new VectorClock(readUint32BE(bytes, 0), readUint32BE(bytes, 4));
  }

  /** Sum a + b (used for deterministic merge ordering) */
  get total() {
    return this.a + this.b;
  }

  /**
   * Return new VectorClock with counter for given role incremented
   * @param {string} localPeerRole - 'A' or 'B'
   * @returns {VectorClock}
   */
  increment(localPeerRole) {
    if (localPeerRole === 'A') return new VectorClock(this.a + 1, this.b);
    if (localPeerRole === 'B') return new VectorClock(this.a, this.b + 1);
    throw new Error(`Invalid peer role: ${localPeerRole}. Must be 'A' or 'B'`);
  }

  /**
   * Component-wise maximum (merge)
   * @param {VectorClock} other
   * @returns {VectorClock}
   */
  merge(other) {
    return new VectorClock(
      Math.max(this.a, other.a),
      Math.max(this.b, other.b)
    );
  }

  /**
   * Causal relationship with another vector clock
   * @param {VectorClock} other
   * @returns {string} CausalRelation value
   */
  causalRelation(other) {
    const thisLeqA = this.a <= other.a;
    const thisLeqB = this.b <= other.b;
    const otherLeqA = other.a <= this.a;
    const otherLeqB = other.b <= this.b;

    const thisLeq = thisLeqA && thisLeqB;
    const otherLeq = otherLeqA && otherLeqB;

    if (thisLeq && otherLeq) return CausalRelation.EQUAL;
    if (thisLeq) return CausalRelation.BEFORE;
    if (otherLeq) return CausalRelation.AFTER;
    return CausalRelation.CONCURRENT;
  }

  toJSON() {
    return { a: this.a, b: this.b };
  }

  /**
   * Serialize to 8 bytes (4 for A, 4 for B, big-endian)
   * @returns {Uint8Array}
   */
  toBytes() {
    return concatBytes(uint32BE(this.a), uint32BE(this.b));
  }

  toString() {
    return `VC(${this.a}, ${this.b})`;
  }

  equals(other) {
    return this.a === other.a && this.b === other.b;
  }
}

/**
 * Determines causal relationships between vector clocks.
 */
export class CausalityChecker {
  compare(a, b) {
    return a.causalRelation(b);
  }

  isAfter(event, reference) {
    return event.causalRelation(reference) === CausalRelation.AFTER;
  }

  isConcurrent(a, b) {
    return a.causalRelation(b) === CausalRelation.CONCURRENT;
  }
}

File: packages/ledger_engine/lib/src/vector_clock.dart
import 'dart:typed_data';

import 'package:meta/meta.dart';

/// Causal relationship between two vector clocks.
enum CausalRelation {
  /// This clock is causally before the other.
  before,

  /// This clock is causally after the other.
  after,

  /// The clocks are concurrent (no causal relationship).
  concurrent,

  /// The clocks are identical.
  equal,
}

/// A 2-element vector clock for the Styx 2-peer system.
@immutable
class VectorClock {
  /// Creates a [VectorClock] with components [a] and [b].
  const VectorClock({required this.a, required this.b});

  /// Creates a zero vector clock.
  const VectorClock.zero() : a = 0, b = 0;

  /// Deserializes from a JSON-compatible map.
  factory VectorClock.fromJson(Map<String, dynamic> json) {
    return VectorClock(a: json['a'] as int, b: json['b'] as int);
  }

  /// Counter for peer A.
  final int a;

  /// Counter for peer B.
  final int b;

  /// Increments the counter for [localPeerRole] ('A' or 'B').
  VectorClock increment(String localPeerRole) {
    return switch (localPeerRole) {
      'A' => VectorClock(a: a + 1, b: b),
      'B' => VectorClock(a: a, b: b + 1),
      _ => throw ArgumentError.value(
        localPeerRole,
        'localPeerRole',
        "Must be 'A' or 'B'",
      ),
    };
  }

  /// Merges with [other] by taking the component-wise maximum.
  VectorClock merge(VectorClock other) {
    return VectorClock(
      a: a > other.a ? a : other.a,
      b: b > other.b ? b : other.b,
    );
  }

  /// Sum of all counters (used for deterministic ordering).
  int get total => a + b;

  /// Compares the causal relationship with [other].
  CausalRelation causalRelation(VectorClock other) {
    if (a == other.a && b == other.b) return CausalRelation.equal;
    if (a <= other.a && b <= other.b) return CausalRelation.before;
    if (a >= other.a && b >= other.b) return CausalRelation.after;
    return CausalRelation.concurrent;
  }

  /// Serializes to a JSON-compatible map.
  Map<String, int> toJson() => {'a': a, 'b': b};

  /// Serializes to bytes (8 bytes: 4 for A, 4 for B, big-endian).
  Uint8List toBytes() {
    final data = ByteData(8)
      ..setInt32(0, a)
      ..setInt32(4, b);
    return data.buffer.asUint8List();
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is VectorClock && a == other.a && b == other.b;

  @override
  int get hashCode => Object.hash(a, b);

  @override
  String toString() => 'VectorClock(a: $a, b: $b)';
}
